### ISSUE TITLE
Missing Inferring of the return type of function and warning the user

### ISSUE THREAD

**Issue by ramvalleru** [intellij-rust/intellij-rust#1772]
The plugin doesn't infer the return type of the function and lint possible missing of the return type of function definition.

I created a function that returns an u32 but haven't declared the return type. This is a compile time error but is there any way the plugin can lint a warning or error for missing return type of the function.

```rust
fn first_word(name : &String) {
  name.len();
}
```
When I miss providing the return type in the function definition, a lint showing missing return type for function or possible suggestion of return type would be useful.

**Comment by matklad** [comment]
A somewhat similar check in the spirit is `org.rust.ide.inspections.RsExtraSemicolonInspection`, which checks for the opposite situation: there's a return type, but the tail expression ends with a semicolon.

**Comment by vlad20012** [comment]
I think it's E-hard. Function may have multiple return points in a different branches, so some kind of control flow analysis needed.

**Comment by matklad** [comment]
Hm, we don't need control flow, we need to determine the return points of the function which is already done because we can highlight them :)

**Comment by vlad20012** [comment]
Oh, this issue description looks really ambiguous... The actual example, as I understand, is:
```rust
fn foo() {
    return 0u32;
}//        ~~~~ type mismatch (and provide quick fix)
```

**Comment by ramvalleru** [comment]
**vlad20012** I made a change to my original question with example.

**Comment by vlad20012** [comment]
@ramvalleru There are no errors in your example =D (and start code with ```rust, so highlighting will be provided)

**Comment by vlad20012** [comment]
Anyway, it should be a part of typecheck to emit E0308 here (I'm working on it), and then it's yet another special case of #1730

**Comment by vlad20012** [comment]
![image](https://user-images.githubusercontent.com/3221931/36351855-80791492-14c0-11e8-89e9-250d97ed8624.png)

I think it's fixed now. For the quick fix there is another issue #1730
